playground/pspin/pspin.py

```diff
@@ -0,0 +1,62 @@
+import matplotlib
+import numpy as np
+import matplotlib.pyplot as plt
+import scipy
+from scipy.misc import comb, factorial
+from pele.optimize._quench import lbfgs_cpp, modifiedfire_cpp
+from pele.potentials import MeanFieldPSpinSpherical
+from joblib import Parallel, delayed
+
+
+p=3
+nspins=20
+interactions = np.ones(np.power(nspins,p))
+coords = np.ones(nspins)
+pot = MeanFieldPSpinSpherical(interactions, nspins, p, tol=1e-6)
+e = pot.getEnergy(coords)
+assert e == -comb(nspins,p)/nspins
+
+interactions = np.random.normal(0, np.sqrt(factorial(p)), [nspins for i in xrange(p)])
+#interactions =  (interactions + interactions.transpose())/2
+for i in xrange(nspins):
+    for j in xrange(i, nspins):
+        for k in xrange(j, nspins):
+            interactions[k][i][j] = interactions[i][j][k]
+            interactions[k][j][i] = interactions[i][j][k]
+            interactions[j][k][i] = interactions[i][j][k]
+            interactions[i][k][j] = interactions[i][j][k]
+            interactions[j][i][k] = interactions[i][j][k]
```

Comment:
yes, so the model is defined so that either each permutation of J_ijk has a different value ~ N(0,1) and you actually sum the p! permutations of spins or the interaction matrix is symmetric with J_ijk ~ N(0,sqrt(p!)) and you pick each combinations only once.

If I symmetrise the way you suggest I would have to sample J_ijk ~ N(0,p!/2) and then do A + A.T which should give samples from the right distribution. I have tried both using A.transpose directly and your suggestion and neither seems to work: the minimization becomes really slow, probably because something goes wrong when the assumption that the matrix is symmetric is broken.

A way to write that loop in a general way, for arbitrary p, is using the python next permutation iterator, which I'll do next. Ultimately I will generate one interaction matrix, store it and import it, so I am not too worried about this step.